airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/cursor/file_based_concurrent_cursor.py

```diff
@@ -0,0 +1,272 @@
+from airbyte_cdk.sources.file_based.stream.concurrent.adapters import FileBasedStreamPartition
+from airbyte_cdk.sources.file_based.stream.concurrent.cursor.abstract_concurrent_file_based_cursor import AbstractConcurrentFileBasedCursor
+from airbyte_cdk.sources.file_based.stream.cursor import DefaultFileBasedCursor
+from airbyte_cdk.sources.file_based.types import StreamState
+from airbyte_cdk.sources.message.repository import MessageRepository
+from airbyte_cdk.sources.streams.concurrent.cursor import CursorField
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+
+
+class FileBasedConcurrentCursor(AbstractConcurrentFileBasedCursor):
+    CURSOR_FIELD = "_ab_source_file_last_modified"
+    DEFAULT_DAYS_TO_SYNC_IF_HISTORY_IS_FULL = DefaultFileBasedCursor.DEFAULT_DAYS_TO_SYNC_IF_HISTORY_IS_FULL
+    DEFAULT_MAX_HISTORY_SIZE = 10_000
+    DATE_TIME_FORMAT = DefaultFileBasedCursor.DATE_TIME_FORMAT
+    zero_value = datetime.min
+
+    def __init__(
+        self,
+        stream_config: FileBasedStreamConfig,
+        stream_name: str,
+        stream_namespace: Optional[str],
+        stream_state: MutableMapping[str, Any],
+        message_repository: MessageRepository,
+        connector_state_manager: ConnectorStateManager,
+        cursor_field: CursorField,
+    ) -> None:
+        self._stream_name = stream_name
+        self._stream_namespace = stream_namespace
+        self._state = stream_state
+        self._message_repository = message_repository
+        self._connector_state_manager = connector_state_manager
+        self._cursor_field = cursor_field
+        self._time_window_if_history_is_full = timedelta(
+            days=stream_config.days_to_sync_if_history_is_full or self.DEFAULT_DAYS_TO_SYNC_IF_HISTORY_IS_FULL
+        )
+        # These should be set before the sync starts
+        self._state_lock = RLock()
+        self._pending_files_lock = RLock()
+        self._pending_files: Optional[Dict[str, FileBasedStreamPartition]] = None
+        self._file_to_datetime_history = None
+        self._prev_cursor_value = None
+        self._sync_start = None
+
+        self._file_to_datetime_history = stream_state.get("history", {}) if stream_state else {}
+        self._prev_cursor_value = self._compute_prev_sync_cursor(stream_state)
+        self._sync_start = self._compute_start_time()
+
+    @property
+    def state(self) -> MutableMapping[str, Any]:
+        return self._state
+
+    def observe(self, record: Record) -> None:
+        pass
+
+    def close_partition(self, partition: FileBasedStreamPartition) -> None:
+        with self._pending_files_lock:
+            if self._pending_files is None:
+                raise RuntimeError("Expected pending partitions to be set but it was not. This is unexpected. Please contact Support.")
+
+    def set_pending_partitions(self, partitions: List[FileBasedStreamPartition]) -> None:
+        self._pending_files = {}
```

Comment:
This is stateful and not thread-safe. Do we have a guarantee somewhere that there is only one thread that will call this?